### PR TITLE
initial

### PR DESCRIPTION
**Pull request by tahackr**

### CODE REVIEW

File: src/App.js [hunk 1]
```diff
@@ -1,7 +1,8 @@
-import React from "react";
+import * as React from "react";
+import Table from "./components/Table";
 
 function App() {
-    return <div>App</div>;
+    return <Table />;
```
Comment (Zoz-trk) [inline src/App.js:5]: - [x] App Component should preserve its default rendering stage as you shown in this example.
- [ ] One improvement could be adding container and wrapping sub components inside this wrapper for every other component that will be rendered on the page so that you could easily follow separation of concerns paradigm.

File: src/components/Table.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+import useFetch from "../hooks/use-fetch";
+
+function Table() {
+    //todos
+    //const renderedRows = data.map((data) => <TableBodyRow data={data} />);
+    // Send response.data here
+    // data maplerken her data objesinde quote.USD ye bak
+    // burda .price .percent_change_1h .percent_change_24h .percent_change_7d
+
+    const data = useFetch("v1/cryptocurrency/listings/latest");
```
Comment (Zoz-trk) [inline src/components/Table.js:13]: - [ ] Destructuring response and renaming response data semantically would help with readability and maintainability of data.

File: src/hooks/use-fetch.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+    (async function () {
+        try {
+            const data = await axios.get(url, {
+                headers: {
+                    "X-CMC_PRO_API_KEY": "ab0cf1ed-2b3d-488f-b84e-515d554e6d8a",
+                },
+            });
+            console.log(data);
+            setCoins(data);
```
Comment (Zoz-trk) [inline src/hooks/use-fetch.js:16]: - [ ] How do you know if data is returned without any issue ?

File: src/hooks/use-fetch.js [hunk 4]
```diff
@@ -0,0 +1,25 @@
+import * as React from "react";
+import axios from "axios";
+
+function useFetch(url) {
+    const [coins, setCoins] = React.useState(null);
+    console.log(url);
+
+    (async function () {
```
Comment (Zoz-trk) [inline src/hooks/use-fetch.js:8]: - [ ] What`s the best practice when fetching the data in React.js ? Should we use immediately invoked functions as you did or something else ?

File: src/hooks/use-fetch.js [hunk 5]
```diff
@@ -0,0 +1,25 @@
+import * as React from "react";
+import axios from "axios";
+
+function useFetch(url) {
+    const [coins, setCoins] = React.useState(null);
+    console.log(url);
```
Comment (Zoz-trk) [inline src/hooks/use-fetch.js:6]: - [ ] Clean all the console.log`s from your PR`s

File: src/hooks/use-fetch.js [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+    (async function () {
+        try {
+            const data = await axios.get(url, {
+                headers: {
```
Comment (Zoz-trk) [inline src/hooks/use-fetch.js:11]: - [ ] Instead of manipulating headers directly and adding keys in your request, please save necessary informations into .env file.

File: src/hooks/use-fetch.js [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+                    "X-CMC_PRO_API_KEY": "ab0cf1ed-2b3d-488f-b84e-515d554e6d8a",
+                },
+            });
+            console.log(data);
+            setCoins(data);
+        } catch (err) {
+            console.error(err.message);
+        }
+    })();
```
Comment (Zoz-trk) [inline src/hooks/use-fetch.js:20]: - [ ] Main purpose of creating useFetch hook is reusability. Do you think we can use this hook anywhere else in our application ?

File: src/hooks/use-fetch.js [hunk 8]
```diff
@@ -0,0 +1,25 @@
+import * as React from "react";
+import axios from "axios";
+
+function useFetch(url) {
+    const [coins, setCoins] = React.useState(null);
```
Comment (Zoz-trk) [inline src/hooks/use-fetch.js:5]: - [ ] What`s your plan to handle loading and error stages ? Currently, I don`t see any loading indicator. Also, there is only a single console log statement for error stage which is anti-pattern in React applications.

File: tailwind.config.js [hunk 9]
```diff
@@ -1,6 +1,6 @@
 /** @type {import('tailwindcss').Config} */
 module.exports = {
-    content: ["./src/**/*.{js,jsx,ts,tsx}"],
+    content: ["./src/**/*.{html,js}"],
```
Comment (Zoz-trk) [inline tailwind.config.js:3]: - [ ] Are you able to see Tailwind styles being applied to the page ?

File: src/components/Table.js [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+    return (
+        <table className="block table-fixed p-16">
```
Comment (Zoz-trk) [inline src/components/Table.js:16]: - [ ] Are you sure you would like to go with Native Table component ? You can try Material Ui and I believe it will be more sustainable in the long rung.

**Review by Zoz-trk — CHANGES_REQUESTED** [review]
- [ ] Please review comments and refactor necessary parts.
- [ ] Separation of concerns are key in React folder structure. Try creating relevant folders or files when necessary.
- [ ] Create json file for mock data to be used in your fetch request so that you wont have to work with static inputs in your components.
- [ ] Ignore api configurations for now until you create first version of UI with mock data.

File: src/components/Table.js (hunk 10 above)
Comment (tahackr) [reply src/components/Table.js:16]: Instead of using a Native Table component, I have implemented a Material UI Table component.

File: tailwind.config.js (hunk 9 above)
Comment (tahackr) [reply tailwind.config.js:3]: Yes, I am.

File: src/hooks/use-fetch.js (hunk 8 above)
Comment (tahackr) [reply src/hooks/use-fetch.js:5]: I have restructured the use-fetch hook in order to handle fetching and error statuses.

File: src/hooks/use-fetch.js (hunk 7 above)
Comment (tahackr) [reply src/hooks/use-fetch.js:20]: I have restructured it to be reusable.

File: src/hooks/use-fetch.js (hunk 6 above)
Comment (tahackr) [reply src/hooks/use-fetch.js:11]: I am having issues with creating an .env file. Webpack version 5 does not support "dotenv" package.Following the instructions Webpack gave, I installed a few more packages and created "resolve.fallback" object in the webpack.config file.Even though I don't get any errors now, importing dotenv/config does not create a process object therefore trying to access it breaks the code.

File: src/hooks/use-fetch.js (hunk 4 above)
Comment (tahackr) [reply src/hooks/use-fetch.js:8]: I have made use of "useEffect" hook and moved my function into it.

File: src/App.js (hunk 1 above)
Comment (tahackr) [reply src/App.js]: I have wrapped the newly implemented Material UI Table component into a div.

File: src/components/Table.js (hunk 2 above)
Comment (tahackr) [reply src/components/Table.js]: I have destructured the data I get from "useFetch" hook. Hook returns an object with data,isFetching,error properties.I am not making use of isFetching and data right now. After fixing the issue with http requests I will make use of the data, for isFetching, I can make a Skeleton component to use during fetching if requested.

File: src/hooks/use-fetch.js (hunk 3 above)
Comment (tahackr) [reply src/hooks/use-fetch.js]: I have wrapped setting the data state into an if block. But I wanted to ask, wouldn't try block throw an error and therefore stopping the execution if http request fails?

File: src/components/Table.js [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+            change: -0.5,
+        },
+    ];
+    //todos
+    //const renderedRows = data.map((data) => <TableBodyRow data={data} />);
+    // data maplerken her data objesinde quote.USD ye bak
+    // burda .price .percent_change_1h .percent_change_24h .percent_change_7d
+
+    const {
+        data: deneme,
```
Comment (Zoz-trk) [inline src/components/Table.js:37]: - [ ] Always make sure that you use semantic naming conventions for your variables. "Deneme" is not one of them.

File: src/components/TableBodyRow.js [hunk 12]
```diff
@@ -0,0 +1,14 @@
+import * as React from "react";
+import { TableRow as TR, TableCell as TD } from "@mui/material";
+
+function TableBodyRow({ data }) {
+    return (
+        <TR className="text-center border-b">
+            <th className="text-left">{data.name}</th>
```
Comment (Zoz-trk) [inline src/components/TableBodyRow.js:7]: - [ ] Its better to destructure data object before you use any properties of it.

File: src/hooks/use-fetch.js [hunk 13]
```diff
@@ -0,0 +1,37 @@
+import * as React from "react";
```
Comment (Zoz-trk) [inline src/hooks/use-fetch.js:1]: - [ ] Directly import necessary React hooks. It will improve readibility.

File: src/index.css [hunk 14]
```diff
@@ -1,3 +1,14 @@
 @tailwind base;
 @tailwind components;
 @tailwind utilities;
+
+body {
```
Comment (Zoz-trk) [inline src/index.css:5]: - [ ] Why do you need styling properties in index.css file ? Either use tailwind as inline styling or prefer Material UI way of doing themes.

**Review by Zoz-trk — CHANGES_REQUESTED** [review]

File: src/components/Table.js [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+    Table as TableMUI,
+    TableHead,
+    TableBody,
+    TableRow,
+    TableCell as TD,
+} from "@mui/material";
+import TableBodyRow from "./TableBodyRow";
+import useFetch from "../hooks/use-fetch";
+
+function Table() {
```
Comment (Zoz-trk) [inline src/components/Table.js:13]: - [ ] You should move mock data into its own folder. It could be a folder under src directory named "data" and under data folder you may have a file like mockData.json.

File: src/hooks/use-fetch.js [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+            try {
+                setIsFetching(true);
+                const data = await axios.get(url, {
+                    headers: {
+                        "X-CMC_PRO_API_KEY":
+                            "ab0cf1ed-2b3d-488f-b84e-515d554e6d8a",
+                    },
+                });
+
+                if (data) {
```
Comment (Zoz-trk) [inline src/hooks/use-fetch.js:21]: - [ ] response data is object and you need a way of verifying if response is ok. Check possible return values of get requests.

**Review by Zoz-trk — CHANGES_REQUESTED** [review]

File: src/hooks/use-fetch.js (hunk 3 above)
Comment (Zoz-trk) [reply src/hooks/use-fetch.js]: Catch would grab most of the errors but there are some others which should be verified right after fetch process. I`d suggest making research on this topic. Server or API`s doesn`t always return proper error states.

File: src/hooks/use-fetch.js (hunk 16 above)
Comment (tahackr) [reply src/hooks/use-fetch.js:21]: Made these changes.

```
const { data, status } = await axios.get(url);

                if (status === 200) {
                    setData(data);
                }
```

File: src/hooks/use-fetch.js (hunk 13 above)
Comment (tahackr) [reply src/hooks/use-fetch.js:1]: I have never imported it this way. Faced an error "React is not defined" when i was creating the extension and tried to import it like this.Fixed the error then.But now just tried to import it directly and I don't get the error anymore. 

TL;DR changed imports

File: src/components/TableBodyRow.js (hunk 12 above)
Comment (tahackr) [reply src/components/TableBodyRow.js:7]: Is it better to destructure the data in the props like this:
`function TableBodyRow({ data: { price, change, name } })`
or like this:
```
function TableBodyRow({ data })
const { price, change, name } = data;
```
The first one is less code but wanted to be sure about the conventions or readability issues if this cause any.

File: src/components/Table.js (hunk 15 above)
Comment (tahackr) [reply src/components/Table.js:13]: I used "json-server" package in my earlier projects to learn more about http requests. Installed it and created a db.json file.

File: src/components/TableBodyRow.js (hunk 12 above)
Comment (Zoz-trk) [reply src/components/TableBodyRow.js]: I would prefer second option.

**Review by Zoz-trk — APPROVED** [review]

**Comment by Zoz-trk** [comment]
Looks good to me therefore merging this PR.
We`ll work on further improvements later on.
Great job!
